trackbase/FairRKPropagator.cxx

```diff
@@ -0,0 +1,685 @@
+
+    if ( pcastruct.PCA != 1 && pcastruct.PCA != 2 ) {
+        LOG(info) << "FairRKPropagator::FindPCA implemented for point (pca=1) and wire (pca=2) only";
+        return 1;
+    }
+    double charge = GetChargeFromPDG(pcastruct.PDGCode);
+    double momIn  = fPCAPropagationDir* // if set to -1, it will back propagate
+        TMath::Sqrt(fPCAPropagationPar->GetPx()*fPCAPropagationPar->GetPx()+
+                    fPCAPropagationPar->GetPy()*fPCAPropagationPar->GetPy()+
+                    fPCAPropagationPar->GetPz()*fPCAPropagationPar->GetPz());
+    if ( momIn == 0. ) return 1;
+    double vecIn[7] = {fPCAPropagationPar->GetX(),        fPCAPropagationPar->GetY(),        fPCAPropagationPar->GetZ(),
+                       fPCAPropagationPar->GetPx()/momIn, fPCAPropagationPar->GetPy()/momIn, fPCAPropagationPar->GetPz()/momIn,
+                       momIn};
+
+    double diff;
+    if ( pcastruct.PCA == 1 )
+        diff = sqrt((vecIn[0]-pcastruct.Point.X())*(vecIn[0]-pcastruct.Point.X())+
+                    (vecIn[1]-pcastruct.Point.Y())*(vecIn[1]-pcastruct.Point.Y())+
+                    (vecIn[2]-pcastruct.Point.Z())*(vecIn[2]-pcastruct.Point.Z()));
+    else //if ( pcastruct.PCA == 2 )
+        diff = CalculatePointToWireDistance(TVector3(fPCAPropagationPar->GetX(), fPCAPropagationPar->GetY(), fPCAPropagationPar->GetZ()), pcastruct.Wire1, pcastruct.Wire2, pcastruct.OnWirePCA);
+
+    fMaxStep = diff/25;
+    double res_old= diff;
+    double res = 100.0;
+    double vecOut[7];
+    double vecOutT[7];
+    for (int i=0; i< 7; i++) {vecOut[i]=0; vecOutT[i]=0;}
+
+    int nIter=0;
+    pcastruct.TrackLength = 0.;
+
+    do {
+        double stepLength = Step(charge,vecIn,vecOut);
+        double newDiff;
+        if ( pcastruct.PCA == 1 )
+            newDiff = sqrt((vecOut[0]-pcastruct.Point.X())*(vecOut[0]-pcastruct.Point.X())+
+                           (vecOut[1]-pcastruct.Point.Y())*(vecOut[1]-pcastruct.Point.Y())+
+                           (vecOut[2]-pcastruct.Point.Z())*(vecOut[2]-pcastruct.Point.Z()));
+        else //if ( pcastruct.PCA == 2 )
+            newDiff = CalculatePointToWireDistance(TVector3(vecOut[0], vecOut[1], vecOut[2]), pcastruct.Wire1, pcastruct.Wire2, pcastruct.OnWirePCA);
+
+        res=newDiff/diff;
+        if( TMath::Abs(res)< 0.01 || res >res_old ) {
+            break;
+        } else {
+            for (int i=0; i< 7; i++) {
+                vecOutT[i]=vecOut[i];
+                vecIn  [i]=vecOut[i];
+            }
+            res_old=res;
+            pcastruct.TrackLength += stepLength;
+        }
+        if(nIter++>1000) { break; }
+    } while(1);
+    if (res > res_old) for (int k=0; k< 7; k++) { vecOut[k]=vecOutT[k]; }
+
+    pcastruct.OnTrackPCA.SetX(vecOut[0]);
+    pcastruct.OnTrackPCA.SetY(vecOut[1]);
+    pcastruct.OnTrackPCA.SetZ(vecOut[2]);
+
+    if ( pcastruct.PCA == 1 )
+        pcastruct.Distance = sqrt((vecOut[0]-pcastruct.Point.X())*(vecOut[0]-pcastruct.Point.X())+
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'pcastruct.Distance' is assigned a value that is never used.](https://app.codacy.com/app/FairRootGroup/FairRoot/pullRequest?prid=4139720)